server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -58,26 +58,15 @@ public DelayedOperation(long delayMs) {
      * true, others will still return false.
      */
     public boolean forceComplete() {
-        // Do not proceed if the operation is already completed.
-        if (completed) {
+        if (!completed) {
+            completed = true;
+            // cancel the timeout timer
+            cancel();
+            onComplete();
+            return true;
```

Comment:
Hmm, this doesn't seem to work for `DelayedFuturePurgatory` since it calls forceComplete() without holding a lock.

`           BiConsumer<Void, Throwable> callbackAction = (result, exception) -> delayedFuture.forceComplete();
`

If there is no benefit from removing the return type of `forceComplete()`, we could probably just dismiss this PR.

Reply:
> If there is no benefit from removing the return type of forceComplete(), we could probably just dismiss this PR.

yes, it seems we can't remove the lock and returned type. +1 to close PR

Reply:
Thanks for review, I will close this PR